### PR TITLE
SoftLayer as Sync Layer

### PR DESCRIPTION
**Pull request by cerw**
HI

this is my first push request on GitHub but i have implemented SoftLayer ObjectStorage sync and it works for me. Woudl you like to merge into your great Application? 

Let me know feedback so I can learn a bit :) 

Thanks

Petr

### CODE REVIEW

File: src/Backup/Sync/SoftLayer.php [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+use phpbu\App\Result;
+use phpbu\Backup\Sync;
+use phpbu\Backup\Target;
+use phpbu\Util\Arr;
+use phpbu\Util\String;
+use ObjectStorage_Http_Client;
+use ObjectStorage;
+
+/**
+ * Amazon S3 Sync
```
Comment (sebastianfeldmann) [inline src/Backup/Sync/SoftLayer.php]: Should be "Softlayer ObjectStorage Sync"

File: src/Backup/Sync/SoftLayer.php [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+            throw new Exception('SoftLayer container name is mandatory');
+        }
+        if (!Arr::isSetAndNotEmptyString($config, 'host')) {
+            throw new Exception('SoftLayer host is mandatory');
+        }
+        if (!Arr::isSetAndNotEmptyString($config, 'path')) {
+            throw new Exception('SoftLayer path / object-key is mandatory');
+        }
+        $this->username    = $config['username'];
+        $this->secret =		 $config['secret'];
```
Comment (sebastianfeldmann) [inline src/Backup/Sync/SoftLayer.php]: Please fix the indentation (no Tabs, equal signs aligned)

File: src/Backup/Sync/SoftLayer.php [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+		// If no adapter option is provided, CURL will be used.
+		$options = array('adapter' => ObjectStorage_Http_Client::SOCKET, 'timeout' => 20);
+		$objectStorage = new ObjectStorage($this->host, $this->username, $this->secret, $options);
+
+
+
+
+        try {
+			$newObject = $objectStorage->with($this->container."/".$targetPath)
```
Comment (sebastianfeldmann) [inline src/Backup/Sync/SoftLayer.php]: Indentation again, please make sure to use spaces to indent.

File: src/App/Factory.php [hunk 4]
```diff
@@ -41,12 +41,13 @@
...
-            'ftp'      => '\\phpbu\\Backup\\Sync\\Ftp',
-            'rsync'    => '\\phpbu\\Backup\\Sync\\Rsync',
-            'sftp'     => '\\phpbu\\Backup\\Sync\\Sftp',
+            'amazons3'  => '\\phpbu\\Backup\\Sync\\AmazonS3',
+            'copycom'   => '\\phpbu\\Backup\\Sync\\Copycom',
+            'dropbox'   => '\\phpbu\\Backup\\Sync\\Dropbox',
+            'ftp'       => '\\phpbu\\Backup\\Sync\\Ftp',
+            'rsync'     => '\\phpbu\\Backup\\Sync\\Rsync',
+            'sftp'      => '\\phpbu\\Backup\\Sync\\Sftp',
+			'softlayer' => '\\phpbu\\Backup\\Sync\\SoftLayer',
```
Comment (sebastianfeldmann) [inline src/App/Factory.php]: Indentation (tabs => spaces)

File: src/Backup/Sync/SoftLayer.php [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+use ObjectStorage_Http_Client;
+use ObjectStorage;
+
+/**
+ * Amazon S3 Sync
+ *
+ * @package    phpbu
+ * @subpackage Backup
+ * @author     Petr Cervenka  <petr@nanosolutions.io>
+ * @copyright  Petr Cervenka  <petr@nanosolutions.io>
```
Comment (sebastianfeldmann) [inline src/Backup/Sync/SoftLayer.php]: I'm not an copyright expert, but to be safe, would it be ok for you to change the copyright to
"Sebastian Feldmann <sebastian@phpbu.de>"

File: src/Backup/Sync/SoftLayer.php [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+ * @license    http://www.opensource.org/licenses/BSD-3-Clause  The BSD 3-Clause License
+ * @link       http://phpbu.de/
+ * @since      Class available since Release ?
```
Comment (sebastianfeldmann) [inline src/Backup/Sync/SoftLayer.php]: Next release ist 1.1.6 so you can use this.
Same for the UnitTest :)

File: tests/phpbu/Backup/Sync/SoftLayerTest.php [hunk 7]
```diff
@@ -0,0 +1,113 @@
+<?php
+namespace phpbu\Backup\Sync;
+
+/**
+ * SoftLayerTest
+ *
+ * @package    phpbu
+ * @subpackage tests
+ * @author     Petr Cervenka  <petr@nanosolutions.io>
+ * @copyright  Petr Cervenka  <petr@nanosolutions.io>
```
Comment (sebastianfeldmann) [inline tests/phpbu/Backup/Sync/SoftLayerTest.php]: Same as 'src/Backup/Sync/SoftLayer.php'

**Comment by sebastianfeldmann** [comment]
Nice job, will merge after some small fixes.
See comments above ;)

**Comment by cerw** [comment]
Will fix up asap thanks

**Comment by cerw** [comment]
Ok let me know if i did well. Why is the CI failing?

**Comment by sebastianfeldmann** [comment]
CI was failing becaus the Code-Coverage-Report did not get there in time, maybe i have to increase the timeouts there. I restarted the build and everything is "green" now ;)

**Comment by sebastianfeldmann** [comment]
*merged*
Looks great, thank you!

**Comment by sebastianfeldmann** [comment]
Hi Petr,
I made some improvements/changes.
You hardcoded the Header MIME Type to 'application/x-bzip2', I added ```Target::getMimeType``` to get the correct MIME Type for a Target configuration.
I renamed 'username' to 'user' for consistency with other 'Sync' classes and I added an example config ```doc/conf/sync/softlayer.xml```

**Comment by cerw** [comment]
Perfect, I thought that would be better way to get the MIME, I will checkout your version and test it out and then roll to my machines.. 

:+1:

**Comment by sebastianfeldmann** [comment]
I currently don't have a testaccount (working on that).
Please let me know if everything is still working, so i can release a new Version.

**Comment by cerw** [comment]
HI
I have just checkotu your master but it seem you have lots of new changes and master seem broken. 

```
PHP Fatal error:  Class 'phpbu\App\Cmd' not found in /Users/cerw/Work/nano/phpbu/phpbu on line 57
```

However the 591563fea451d47b70e5147e9357fbb93ce99d55 works fine.

**Comment by sebastianfeldmann** [comment]
Please run 
    composer update
to update the autoloader

**Comment by cerw** [comment]
Exception 'phpbu\App\Exception' with message 'unknown sync: softlayer'
in phar:///usr/local/bin/phpbu/App/Factory.php:73

**Comment by sebastianfeldmann** [comment]
OK i'm getting confused.
The PHAR version is not yet updated, so the *softlayer* sync is not available there.
You have to use a *clone* or the *dev-master* via composer to check the *softlayer* functionality.

But as you said 591563fea451d47b70e5147e9357fbb93ce99d55 worked fine so i will merge the *softlayer* integration into the next version.

Will release a new stable PHAR version with *softlayer* support hopefully this week.

Thanks again :)

**Comment by cerw** [comment]
OK my bed

 ./phpbu --configuration=/tmp/cerw.xml 

work without ./ it was using old PHAR version.

**Comment by sebastianfeldmann** [comment]
Awesome!

**Comment by sebastianfeldmann** [comment]
@cerw 
If you work with a clone i recommend to remove

    Line 20:    "minimum-stability": "dev",

from your local *composer.json* before running *composer install/update* so you don't use all the *dev-packages* but only those only available in *dev-versions* like:
- phpseclib
- softlayer-objectstorage

phpseclib is working hard to release a stable version.
Softlayer is a different story, there is an issue to release a stabe version but so far the project seems to be a bit deserted.

**Comment by cerw** [comment]
OK will do, can you please share the way you compile the PHAR? Woudl like to do it my self too. 

petr

**Comment by sebastianfeldmann** [comment]
All code used to generate the PHAR is included in the repository.

I use *ant* to execute the neccessary tasks. You can find all the details in *build.xml*
In short:
- Copy the framework and all dependencies in one directory
- Versioning some files
- Patching some files
- Use *phpab* (https://github.com/theseer/Autoload) to create the PHAR

if you have *ant* and *phpab* installed, just run:

    ant phar

**Pull request merged**